src/stim/io/measure_record_reader.cc

```diff
@@ -415,124 +419,95 @@ void MeasureRecordReaderFormatR8::buffer_data() {
+      buffer(n_measurements + n_detection_events + n_logical_observables),
+      position_in_buffer(n_measurements + n_detection_events + n_logical_observables),
+      m_bits_per_record(n_measurements),
+      d_bits_per_record(n_detection_events),
+      l_bits_per_record(n_logical_observables) {
 }
 
 bool MeasureRecordReaderFormatDets::read_bit() {
-    if (position_m >= bits_per_m_record && position_d >= bits_per_d_record && position_l >= bits_per_l_record) {
-        throw std::out_of_range("Attempt to read past end-of-record");
-    }
-
-    if (position() >= bits_per_record()) {
-        result_type = next_shot_result_type;
+    if (position_in_buffer >= m_bits_per_record + d_bits_per_record + l_bits_per_record) {
+        throw std::invalid_argument("Read past end of buffer.");
     }
-    bool r = next_shot == position()++;
-    if (no_next_shot) {
-        r = false;
-    }
-    if ((no_next_shot || position() > next_shot) && result_type == next_shot_result_type && separator == ' ') {
-        update_next_shot();
-    }
-    return r;
+    return buffer[position_in_buffer++];
 }
 
 bool MeasureRecordReaderFormatDets::next_record() {
-    while (separator == ' ') {
-        update_next_shot();
-    }
-    if (!maybe_consume_keyword(in, "shot", separator)) {
+    return start_record();
+}
+
+bool MeasureRecordReaderFormatDets::start_record() {
+    int c;
+    if (!maybe_consume_keyword(in, "shot", c)) {
         return false;
     }
-    no_next_shot = true;
-    next_shot = 0;
-    position_m = position_d = position_l = 0;
-    update_next_shot();
+    buffer.clear();
+    position_in_buffer = 0;
+    while (true) {
+        bool had_spacing = c == ' ';
+        while (c == ' ') {
+            c = getc(in);
+        }
+        if (c == '\n' || c == EOF) {
+            break;
+        }
+        if (!had_spacing) {
+            throw std::invalid_argument("DETS values must be separated by spaces.");
+        }
+        size_t offset;
+        size_t size;
+        char prefix = c;
+        if (prefix == 'M') {
+            offset = 0;
+            size = m_bits_per_record;
+        } else if (prefix == 'D') {
+            offset = m_bits_per_record;
+            size = d_bits_per_record;
+        } else if (prefix == 'L') {
+            offset = m_bits_per_record + d_bits_per_record;
+            size = l_bits_per_record;
+        } else {
+            throw std::invalid_argument("Unrecognized DETS prefix: '" + std::to_string(c) + "'");
+        }
+        uint64_t number;
+        if (!read_uint64(in, number, c)) {
+            throw std::invalid_argument("DETS prefix '" + std::to_string(c) + "' wasn't not followed by an integer.");
+        }
+        if (number >= size) {
+            throw std::invalid_argument(
+                "Got '" + std::to_string(c) + std::to_string(number) + "' but expected num values of that type is " +
+                std::to_string(size) + ".");
+        }
+        buffer[offset + number] ^= true;
```

Comment:
Plain and simple file formats beat complicated file formats. AFAICT, the complexity introduced by xoring here is unnecessary and will undoubtedly surprise human users. Can we treat duplicate shots as errors instead?

Reply:
Suppose that every `OBSERVABLE_INCLUDE(0)` instruction resulted in an L0 being written to the file or not. So they were accumulated "on disk" instead of in memory. This would be a not-so-completely-crazy design, with a few performance benefits, and I didn't want to block that sort of thing off.

And again; robustness principle. It's very hard for us to enforce these kinds of constraints if other people start writing writers.